internal/session.go

```diff
@@ -0,0 +1,421 @@
+//    if err != nil {
+//        // Session has failed or activity itself failed.
+//    }
+//    ... // execute more activities using sessionCtx
+//    err = CompleteSession(sessionCtx)
+//    if err != nil {
+//        // Wrong ctx is used or failed to release session resource.
+//    }
+func CreateSession(ctx Context) (Context, error) {
+	options := getActivityOptions(ctx)
+	baseTasklist := options.TaskListName
+	if baseTasklist == "" {
+		baseTasklist = options.OriginalTaskListName
+	}
+	return createSession(ctx, getCreationTasklist(baseTasklist), true)
+}
+
+// RecreateSession recreate a session based on the sessionInfo passed in. Activities executed within
+// the recreated session will be executed by the same worker as the previous session. RecreateSession()
+// returns an error under the same situation as CreateSession() and has the same usage as CreateSession().
+// It will not check the state of the session described by the sessionInfo passed in, so user can recreate
+// a session based on a failed or completed session.
+//
+// The main usage of RecreateSession is for long sessions that are splited into multiple runs. At the end of
+// one run, complete the current session, get sessionInfo from the context and pass the information to the
+// next run. In the new run, the session can be recreated on the information.
+func RecreateSession(ctx Context, sessionInfo *SessionInfo) (Context, error) {
+	return createSession(ctx, sessionInfo.tasklist, false)
+}
+
+// CompleteSession completes a session. It releases worker resources, so other sessions can be created.
+//
+// After a session is completed, user can continue to use the context, but the activities will be scheduled
+// on the normal taskList (as user specified in ActivityOptions) and may be picked up by another worker since
+// it's not in a session.
+//
+// This API will return an error if the session in the context has already completed, or the resource release
+// process has failed. No error will returned if user tries to complete a session that has already failed and
+// the API call won't do anything.
+func CompleteSession(ctx Context) error {
+	sessionInfo := getSessionInfo(ctx)
+	if sessionInfo == nil || sessionInfo.sessionState == sessionStateClosed {
+		return errNoOpenSession
+	}
+
+	if sessionInfo.sessionState == sessionStateFailed {
+		return nil
+	}
+	retryPolicy := &RetryPolicy{
+		InitialInterval:    time.Second,
+		BackoffCoefficient: 1.1,
+		MaximumInterval:    time.Second * 10,
+		MaximumAttempts:    5,
+	}
+	ao := ActivityOptions{
+		ScheduleToStartTimeout: time.Second * 10,
+		StartToCloseTimeout:    time.Second * 10,
+		RetryPolicy:            retryPolicy,
+	}
+	completionCtx := WithActivityOptions(ctx, ao)
+	// the tasklist will be overrided to use the one stored in sessionInfo
+	err := ExecuteActivity(completionCtx, sessionCompletionActivityName, sessionInfo.SessionID).Get(ctx, nil)
```

Comment:
I would first cancel sessionCreationActivity to ensure that even the sessionCompletion fails the creationActivity exits on the next heartbeat. 
Then I would not call Get on the Future to not block the workflow on the completion.

Reply:
Done.